spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/repository/query/GqlDatastoreQuery.java

```diff
@@ -112,26 +115,105 @@ public GqlDatastoreQuery(Class<T> type, DatastoreQueryMethod queryMethod,
 	@Override
 	public Object execute(Object[] parameters) {
 		GqlQuery query = bindArgsToGqlQuery(this.gql, getParamTags(), parameters);
-		Iterable found = this.datastoreTemplate.queryKeysOrEntities(query, this.entityType);
+		boolean returnTypeIsCollection = this.queryMethod.isCollectionQuery();
+		Class returnedItemType = returnTypeIsCollection
+				? this.queryMethod.getResultProcessor().getReturnedType()
+						.getReturnedType()
+				: this.queryMethod.getReturnedObjectType();
+
+		return isNonEntityReturnedType(returnedItemType)
+				? executeNonEntityQuery(query, returnTypeIsCollection, returnedItemType)
+				: executeEntityQuery(query, returnTypeIsCollection);
+	}
+
+	private Object executeEntityQuery(GqlQuery query, boolean returnTypeIsCollection) {
+		Iterable found = this.datastoreTemplate.queryKeysOrEntities(query,
+				this.entityType);
+
 		List rawResult = found == null ? Collections.emptyList()
 				: (List) StreamSupport.stream(found.spliterator(), false)
 						.collect(Collectors.toList());
 		Object result;
+
 		if (this.queryMethod.isCountQuery()) {
 			result = rawResult.size();
 		}
 		else if (this.queryMethod.isExistsQuery()) {
 			result = !rawResult.isEmpty();
 		}
-		else if (rawResult.isEmpty() || rawResult.get(0).getClass() == Key.class) {
-			result = rawResult;
+		else if (!returnTypeIsCollection) {
+			return rawResult.isEmpty() ? null
+					: this.queryMethod.getResultProcessor()
+							.processResult(rawResult.get(0));
+		}
+		else {
+			result = this.datastoreTemplate.getDatastoreEntityConverter().getConversions()
+					.convertCollection(applyProjection(rawResult),
+							this.queryMethod.getReturnedObjectType());
+		}
+		return result;
+	}
+
+	private Object executeNonEntityQuery(GqlQuery query, boolean returnTypeIsCollection,
+			Class returnedItemType) {
+		Iterable found = this.datastoreTemplate.query(query, x -> {
+			Object mappedResult;
+			if (x instanceof Key) {
+				mappedResult = x;
+			}
+			else {
+				BaseEntity entity = (BaseEntity) x;
+				mappedResult = entity.getValue((String) entity.getNames().toArray()[0])
+						.get();
+			}
+			return mappedResult;
+		});
+
+		List rawResult = found == null ? Collections.emptyList()
+				: (List) StreamSupport.stream(found.spliterator(), false)
+						.collect(Collectors.toList());
+		Object result;
+
+		if (this.queryMethod.isCountQuery()) {
+			result = rawResult.size();
+		}
+		else if (this.queryMethod.isExistsQuery()) {
+			result = !rawResult.isEmpty();
 		}
 		else {
-			result = applyProjection(rawResult);
+			if (returnTypeIsCollection) {
```

Comment:
Conceptually, this is very similar to TwoStepsConversions.convertOnRead - please take a look if you can reuse it here.

Reply:
@dmitry-s now using the same 2-step convertOnRead function everywhere! PTAL

Reply:
@meltsufin PTAL